### PR TITLE
Add improved performance on tons of whitespace

### PR DESCRIPTION
**Pull request by 42shadow42**
### Initial checklist

*   [x] I read the support docs 
*   [x] I read the contributing guide 
*   [x] I agree to follow the code of conduct 
*   [x] I searched issues and couldn’t find anything (or linked relevant results below) 
*   [x] If applicable, I’ve added docs and tests

### Description of changes

Altered method of clearing out extraneous whitespace to handle non-matches more effectively. Approaches the problem using a regex with no backtracking to remove trailing whitespace, reverses the string and repeats to remove leading (now trailing) whitespace again. reverses the string to restore the original orientation.

I know it seems a little unorthodox to reverse the string, but because of how regex engines operate a leading whitespace regex like "[ \t]*$" runs inefficiently with lots of non-matching whitespace.

### CODE REVIEW

**Comment by 42shadow42** [comment]
Fixes https://github.com/syntax-tree/mdast-util-to-hast/issues/61

**Comment by wooorm** [comment]
Just moving some of the discussion from https://github.com/wooorm/trim-lines/commit/51434f7864f909e0469883dd23add10162f9d5f2#commitcomment-76990903 and https://github.com/wooorm/trim-lines/commit/321dc64468940f6fb59bacc775f5e7e2a940a51e#commitcomment-76991050 back here.

I remembered that I actually had a tiny utility that does what is also done here, that had the same performance problem, so I decided that it was probably a good idea to fix that utility, and reuse it here.

I indeed thought that it would probably be really slow to revere strings in some cases. And it looked a bit off. That’s why I came up with a regex, splitting on line endings, and then trimming each line.

I tested, with your test, that this new regex-way works exactly as fast as your reverse-string-way. You’re right that it would be good to also test some whitespace around line endings!

File: lib/handlers/text.js [hunk 1]
```diff
@@ -12,6 +12,17 @@ import {u} from 'unist-builder'
...
+      String(node.value)
+        .replace(/^[ \t]+/gm, '')
+        .split('')
+        .reverse()
+        .join('')
+        .replace(/^[ \t]+/gm, '')
+        .split('')
+        .reverse()
+        .join('')
+    )
```
Comment (wooorm) [inline lib/handlers/text.js:26]:
```suggestion
    u('text', trimLines(node.value))
```

Could you switch this PR to use `trim-lines` from `npm` as a dependency?

File: test/text.js [hunk 2]
```diff
@@ -63,3 +63,20 @@ test('Nodes', (t) => {
 
   t.end()
 })
+
+test('Nodes Efficiency', (t) => {
```
Comment (wooorm) [inline test/text.js:67]:
```suggestion
test('efficiency', (t) => {
```

File: test/text.js [hunk 3]
```diff
@@ -63,3 +63,20 @@ test('Nodes', (t) => {
...
+  const timeout = setTimeout(() => {
+    t.fail('should process lots of whitespace efficiently')
```
Comment (wooorm) [inline test/text.js:69]:
```suggestion
    t.fail('should process very fast')
```

File: test/text.js [hunk 4]
```diff
@@ -63,3 +63,20 @@ test('Nodes', (t) => {
...
+  }, 10)
+
+  t.deepEqual(
+    toHast(u('text', `1${' '.repeat(70_000)}2`)),
+    u('text', `1${' '.repeat(70_000)}2`),
+    'should be efficient on excessive whitespace'
+  )
```
Comment (wooorm) [inline test/text.js:76]:
```suggestion
  const whitespace = ' '.repeat(70_000)

  t.deepEqual(
    toHast(u('text', '1' + whitespace + '2')),
    u('text', '1' + whitespace + '2'),
    'should be efficient on excessive whitespace'
  )
```

File: test/text.js [hunk 5]
```diff
@@ -63,3 +63,20 @@ test('Nodes', (t) => {
 
   t.end()
 })
+
```
Comment (wooorm) [inline test/text.js:66]: Below are some ideas on cleaning this code.
Optionally: it can also be dropped, as this test (and more) are now in `trim-lines`!

File: lib/handlers/text.js (hunk 1 above)
Comment (42shadow42) [reply lib/handlers/text.js:26]: Yes, I think it makes sense to utilize an equivalent library that can be independently maintained.

File: test/text.js (hunk 5 above)
Comment (42shadow42) [reply test/text.js:66]: I think I'll just drop this test, I'll make the change once the library has the final revision in place. See my PR for details. https://github.com/wooorm/trim-lines/pull/5

Comment (wooorm) [reply test/text.js:66]: sounds good

**Comment by 42shadow42** [comment]
@wooorm This one is ready for rereview. It now just delegates the inefficiency to your trim-lines package.

File: package.json [hunk 6]
```diff
@@ -41,6 +41,7 @@
     "mdast-util-definitions": "^5.0.0",
     "mdurl": "^1.0.0",
     "micromark-util-sanitize-uri": "^1.0.0",
+    "trim-lines": "^3.0.1",
```
Comment (wooorm) [inline package.json:44]:
```suggestion
    "trim-lines": "^3.0.0",
```

Using loose semver ranges makes it easier for package managers to install things and keeps lock file diffs smaller.

Comment (wooorm) [reply package.json:44]: Semver solves this.

**Review by wooorm — APPROVED** [review]

**Comment by wooorm** [comment]
Thanks, released in [12.1.2](https://github.com/syntax-tree/mdast-util-to-hast/releases/tag/12.1.2)!
